src/routes/account.js

```diff
@@ -407,5 +407,138 @@ router.post('/stripe/webhook', async (request, response) => {
 	response.json({ received: true });
 });
 
+router.get('/sso/discourse', async (request, response, next) => {
+	if (!request.query.sso || !request.query.sig) {
+		return next(); // * 404
+	}
+
+	const signature = util.signDiscoursePayload(request.query.sso);
+
+	if (signature !== request.query.sig) {
+		return next(); // * 404
+	}
+
+	const decodedPayload = new URLSearchParams(Buffer.from(request.query.sso, 'base64').toString());
+
+	if (!decodedPayload.has('nonce') || !decodedPayload.has('return_sso_url')) {
+		return next(); // * 404
+	}
+
+	// * User already logged in, don't show the login prompt
+	if (request.cookies.access_token && request.cookies.refresh_token) {
+		try {
+			const accountData = await util.getUserAccountData(request, response);
+
+			// * Discourse REQUIRES unique emails, however we do not due to NN also
+			// * not requiring unique email addresses. Email addresses, for now,
+			// * are faked using the users PID. This will essentially disable email
+			// * for the forum, but it's a bullet we have to bite for right now.
+			// TODO - We can run our own SMTP server which maps fake emails (pid@pretendo.whatever) to users real emails
+			const payload = Buffer.from(new URLSearchParams({
+				nonce: decodedPayload.get('nonce'),
+				external_id: accountData.pid,
+				email: `${accountData.pid}@invalid.com`, // * Hack to get unique emails
```

Comment:
If we do ever do `<pid>@pnid.pretendo.network` (or w/e), is it going to be a pain to migrate the existing accounts tied to invalid.com addresses?
Could this be a security risk if the owner of invalid.com were to start requesting password resets?

Reply:
These invalid emails only exist for the Discourse user profiles, and we can mass update those once we have an SMTP server.

There is no password resets because this uses DiscourseConnect, which replaces the original Discourse login flow with one from an external service (in this case, PNID login on the website).

Using `whatever@invalid.com` was the official recommendation from Discourse to handle this situation ([relevant comment](https://meta.discourse.org/t/integration-into-custom-auth-system-where-emails-are-not-unique/306489/28?u=jonbarrow)).

Reply:
Oh wow, I didn't realise how in-depth they went with you. LGTM then